File: apps/windows/TeslaSync.App/feature-views/CostPerKwhChart/CostPerKwhChart.Model.cs
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The mutually-exclusive render branch of the <c>CostPerKwhChart</c> surface — the native union of the
/// states the web component renders
/// (web/src/features/charging/components/cost-analysis/CostPerKwhChart.tsx). The web source is a pure
/// presentational component (it takes a single <c>data: { date; costPerKwh }[]</c> prop and performs no
/// fetching), so the branches are a direct function of the input <see cref="CostPerKwhChartModel"/> — there
/// is no fetch-driven error / stale / offline branch to reproduce here. The parent cost-analysis page owns
/// the query lifecycle (loading / error / stale / offline are handled once for the page before any chart is
/// shown), exactly as the web page only renders this chart once the monthly cost series has resolved. Every
/// branch maps onto a visible surface; none is ever hidden.
/// </summary>
public enum CostPerKwhChartState
{
    /// <summary>Initial fetch in flight (the parent is still loading the series) — title + skeleton chrome.</summary>
    Loading,

    /// <summary>Resolved with no points to plot (web <c>data.length === 0</c>) — a friendly empty state.</summary>
    Empty,

    /// <summary>At least one point is present (web <c>data.length &gt; 0</c>) — the trend line chart.</summary>
    Ready,
}

/// <summary>
/// One blended cost-per-kWh sample — the native mirror of the web datum shape
/// (<c>{ date: string; costPerKwh: number }</c>) the web <c>CostPerKwhChart</c> receives in its
/// <c>data</c> prop. <see cref="Date"/> is the period label the web plots on the <c>XAxis dataKey="date"</c>;
/// <see cref="CostPerKwh"/> is the blended rate in dollars per kWh (the scalar value the web feeds the
/// <c>costPerKwh</c> line, formatted as currency at the display boundary). Pure data — no WinUI types.
/// </summary>
public sealed record CostPerKwhPoint(string Date, double CostPerKwh);

/// <summary>
/// The render-time data model the <c>CostPerKwhChart</c> view binds to — the native analogue of the web
/// <c>CostPerKwhChartProps</c> (<c>data: { date; costPerKwh }[]</c>), plus the fetch flag the parent supplies
/// so the surface can render its own loading branch (the web parent gates the whole chart behind its query
/// state). The component is presentational: this model carries only the sample series and the loading flag.
/// User-facing labels are resolved from the i18n facade by the projection, not passed in. Pure data — no
/// WinUI types — so the projection is unit-tested without a UI host.
/// </summary>
public sealed record CostPerKwhChartModel(bool Loading, IReadOnlyList<CostPerKwhPoint> Points)
{
    /// <summary>The initial model: the first fetch is in flight and no points have arrived yet.</summary>
    public static CostPerKwhChartModel Pending { get; } =
        new(true, Array.Empty<CostPerKwhPoint>());

    /// <summary>A resolved model with no points — the empty state.</summary>
    public static CostPerKwhChartModel Empty { get; } =
        new(false, Array.Empty<CostPerKwhPoint>());
}

/// <summary>
/// One projected, render-ready point on the trend line — the native analogue of a single recharts
/// <c>Line</c> datum. <see cref="Index"/> is the ordinal x-position (the native cartesian surface plots on a
/// numeric domain and surfaces <see cref="DateLabel"/> in the cursor tooltip, the way the web
/// <c>XAxis dataKey="date"</c> labels each tick); <see cref="DateLabel"/> is the period label; and
/// <see cref="CostPerKwh"/> is the blended rate in dollars per kWh plotted on the value axis. Pure data.
/// </summary>
public sealed record CostPerKwhTrendPoint(int Index, string DateLabel, double CostPerKwh);

/// <summary>
/// The fully projected, render-ready view of the chart for one input model — the native analogue of what the
/// web <c>CostPerKwhChart</c> renders inside its <c>GlassPanel</c>. Holds the resolved title (the web
/// <c>t('costAnalysis.charts.costPerKwh')</c> heading), the rate-series label (the web <c>Line name</c>
/// <c>t('costAnalysis.charts.rateLabel')</c>), the empty + loading copy, the active <see cref="State"/>, the
/// projected <see cref="Points"/>, a currency-formatted <see cref="RangeSummary"/> (the spoken analogue of
/// the web's <c>formatCurrency</c> value axis), and the surface <see cref="AutomationName"/>. Pure data so
/// every branch is asserted headlessly.
/// </summary>
public sealed record CostPerKwhChartDisplay(
    CostPerKwhChartState State,
    string Title,
    string RateLabel,
    string EmptyMessage,
    string LoadingLabel,
    IReadOnlyList<CostPerKwhTrendPoint> Points,
    string RangeSummary,
    string AutomationName);

/// <summary>
/// Pure projection from a <see cref="CostPerKwhChartModel"/> to its <see cref="CostPerKwhChartDisplay"/> —
/// the native port of web/src/features/charging/components/cost-analysis/CostPerKwhChart.tsx. The web source
/// is a bare titled line chart of the blended cost-per-kWh trend; this projection reproduces its two
/// <c>t(...)</c> labels (<c>costAnalysis.charts.costPerKwh</c> / <c>costAnalysis.charts.rateLabel</c>), its
/// empty copy (<c>costAnalysis.charts.noData</c>), and the web's <c>data.length &gt; 0</c> gate. The web
/// formats every value-axis tick through <c>useFormatting().formatCurrency(v, 2)</c>; because the native
/// cartesian surface draws a numeric value axis, that currency formatting is reproduced for assistive
/// technology in <see cref="CostPerKwhChartDisplay.RangeSummary"/> via <see cref="ScalarFormatters"/> (the
/// shared SI-scalar currency formatter, the native <c>formatCurrency</c>). Every label resolves through the
/// i18n facade using the same keys the web source feeds into <c>t(...)</c>. No WinUI types — unit-tested
/// without a UI host.
/// </summary>
public static class CostPerKwhChartProjection
{
    private const string EmDash = "\u2014";
    private const string EnDash = "\u2013";
    private const string CurrencySymbol = "$";
    private const int CurrencyDecimals = 2;

    /// <summary>Project <paramref name="model"/> into a render-ready display using the i18n facade.</summary>
    /// <param name="model">The render-time data model (the web <c>data</c> prop + the parent's fetch flag).</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    public static CostPerKwhChartDisplay Project(CostPerKwhChartModel model, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        string title = localizer.GetString("costAnalysis.charts.costPerKwh", "Cost per kWh Trend");
        string rateLabel = localizer.GetString("costAnalysis.charts.rateLabel", "$/kWh");
        string emptyMessage = localizer.GetString("costAnalysis.charts.noData", "Not enough data");
        string loadingLabel = localizer.GetString("common.loading", "Loading");

        IReadOnlyList<CostPerKwhTrendPoint> points = BuildPoints(model.Points);
        CostPerKwhChartState state = SelectState(model.Loading, points);
        string rangeSummary = BuildRangeSummary(points);

        return new CostPerKwhChartDisplay(
            State: state,
            Title: title,
            RateLabel: rateLabel,
            EmptyMessage: emptyMessage,
            LoadingLabel: loadingLabel,
            Points: points,
            RangeSummary: rangeSummary,
            AutomationName: BuildAutomationName(state, title, rateLabel, emptyMessage, loadingLabel, rangeSummary));
    }

    /// <summary>Branch precedence from the web source's data lifecycle: loading → empty → ready.</summary>
    private static CostPerKwhChartState SelectState(bool loading, IReadOnlyList<CostPerKwhTrendPoint> points)
    {
        if (loading)
        {
            return CostPerKwhChartState.Loading;
        }

        // Web gate: `data.length > 0 ? <chart/> : <empty/>`. A single sample is enough to draw the line.
        return points.Count > 0 ? CostPerKwhChartState.Ready : CostPerKwhChartState.Empty;
    }

    private static IReadOnlyList<CostPerKwhTrendPoint> BuildPoints(IReadOnlyList<CostPerKwhPoint> source)
    {
        if (source.Count == 0)
        {
            return Array.Empty<CostPerKwhTrendPoint>();
        }

        var points = new List<CostPerKwhTrendPoint>(source.Count);
        for (int i = 0; i < source.Count; i++)
        {
            var point = source[i];
            string dateLabel = string.IsNullOrWhiteSpace(point.Date) ? EmDash : point.Date;
            double cost = double.IsNaN(point.CostPerKwh) || double.IsInfinity(point.CostPerKwh)
                ? 0.0
                : point.CostPerKwh;
            points.Add(new CostPerKwhTrendPoint(i, dateLabel, cost));
        }

        return points;
    }

    /// <summary>
    /// The spoken, currency-formatted span of the plotted rates — the assistive-technology analogue of the
    /// web's <c>formatCurrency</c> value axis. Empty when there is nothing to plot; a single formatted value
    /// when every sample rounds the same; otherwise "min–max" through the shared SI currency formatter.
    /// </summary>
    private static string BuildRangeSummary(IReadOnlyList<CostPerKwhTrendPoint> points)
    {
        if (points.Count == 0)
        {
            return string.Empty;
        }

        double min = double.PositiveInfinity;
        double max = double.NegativeInfinity;
        foreach (var point in points)
        {
            min = Math.Min(min, point.CostPerKwh);
            max = Math.Max(max, point.CostPerKwh);
        }

        string minText = ScalarFormatters.FormatCurrency(min, CurrencySymbol, CurrencyDecimals);
        string maxText = ScalarFormatters.FormatCurrency(max, CurrencySymbol, CurrencyDecimals);
        return string.Equals(minText, maxText, StringComparison.Ordinal) ? minText : $"{minText}{EnDash}{maxText}";
    }

    private static string BuildAutomationName(
        CostPerKwhChartState state,
        string title,
        string rateLabel,
        string emptyMessage,
        string loadingLabel,
        string rangeSummary) => state switch
        {
            CostPerKwhChartState.Loading => $"{title}. {loadingLabel}",
            CostPerKwhChartState.Empty => $"{title}. {emptyMessage}",
            _ => $"{title}. {rateLabel} {rangeSummary}",
        };
}

/// <summary>
/// PII-safe diagnostics for the <c>CostPerKwhChart</c> surface (P1/S11 diagnostics contract). Records only
/// the operational <c>view.opened</c> event with the surface slug — never a date label or rate value — so a
/// diagnostics line can never leak a user's charging cost. Thread-safe.
/// </summary>
public sealed class CostPerKwhChartDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public CostPerKwhChartDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=CostPerKwhChart</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={CostPerKwhChartRegistration.Slug}");
    }
}

/// <summary>
/// Canonical metadata for the <c>CostPerKwhChart</c> feature surface — the native mirror of the web
/// component at <c>web/src/features/charging/components/cost-analysis/CostPerKwhChart.tsx</c>.
/// </summary>
public static class CostPerKwhChartRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event.</summary>
    public const string Slug = "CostPerKwhChart";
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/CostPerKwhChart/CostPerKwhChartModel.kt
// Pure, framework-free model + projection for the Cost per kWh Trend chart feature view — the native
// analogue of everything the web component needs before it returns JSX
// (web/src/features/charging/components/cost-analysis/CostPerKwhChart.tsx). No Compose, no Android, no
// HTTP: every declaration here is exercised off-device by the :android:testReleaseUnitTest gate, so the
// composable stays a thin render layer over these pure functions.
//
// The web component is purely presentational. Its parent (the Cost Analysis page, via
// `useCostAnalysisData`) builds the `{ date, costPerKwh }[]` series — sorting sessions by start time and
// dividing each session's `cost_decimal` by its energy in kWh — and passes it down. The component then
// renders a single Recharts `<Line dataKey="costPerKwh" name="$/kWh" stroke={palette[2]} />` over a
// `date` X axis with a currency-formatted Y axis (`formatCurrency(v, 2)`), falling back to a
// "Not enough data" message when the series is empty (`data.length > 0`). This file owns that render-ready
// projection plus the `useFormatting` currency contract (`currencySymbol + fmtNumber(amount, 2)`) and the
// accessible-table rows; the composable only resolves localized strings, the palette color, and the
// lifecycle/freshness chrome the shared P1/S8 state layer implies.
//
// SI boundary (ADR / unit-conversion instructions): `costPerKwh` is a money-per-energy ratio, not a
// unit-suffixed physical quantity, so there is no SI conversion here — only the user's display
// `currency_symbol` and decimal rendering, exactly as the web `useFormatting` applies them.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/CostPerKwhChart — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.costperkwhchart

import io.teslasync.android.components.charts.ChartFormat
import io.teslasync.android.data.UiPhase
import io.teslasync.android.data.UiState
import io.teslasync.shared.core.diagnostics.Logger
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import java.util.Locale

/** Default currency symbol when the settings document has none (web `useFormatting` `'$'`). */
internal const val DEFAULT_CURRENCY: String = "$"

/** Cost fraction digits — the web `formatCurrency(v, 2)` literal precision on the Y axis + table. */
internal const val COST_DECIMALS: Int = 2

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object CostPerKwhChartRegistration {
    /** Stable surface id. */
    const val ID: String = "cost-per-kwh-chart"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "CostPerKwhChart"
}

/**
 * One trend point — the native mirror of a web `data` element
 * (`{ date: string; costPerKwh: number }`). [date] is the already-formatted short date the parent emits
 * (the chart X label, an opaque string here exactly as the web `<XAxis dataKey="date" />`), and
 * [costPerKwh] is the session's cost divided by its energy in kWh (the Y value).
 */
data class CostPerKwhPoint(
    val date: String,
    val costPerKwh: Double,
)

/**
 * The user's currency display preference this surface needs — the native port of the web `useFormatting`
 * read of the `/settings` document. Only the [currencySymbol] is needed: the Y axis and the accessible
 * table format with the literal 2-digit precision (web `formatCurrency(v, 2)`), so the user's
 * `decimal_precision` does not apply here.
 */
data class CostCurrencyPrefs(
    val currencySymbol: String,
) {
    companion object {
        /** The `$` default used before settings load (matches the web default). */
        val DEFAULT: CostCurrencyPrefs = CostCurrencyPrefs(DEFAULT_CURRENCY)

        private const val KEY_CURRENCY_SYMBOL = "currency_symbol"

        /** Resolves the currency symbol from the raw `/settings` document (web `useFormatting`). */
        fun fromSettings(settings: JsonElement?): CostCurrencyPrefs {
            val raw = (settings as? JsonObject)?.get(KEY_CURRENCY_SYMBOL) as? JsonPrimitive
            val symbol = raw?.contentOrNull?.trim()
            return CostCurrencyPrefs(if (!symbol.isNullOrEmpty()) symbol else DEFAULT_CURRENCY)
        }
    }
}

/**
 * The already-localized chart microcopy the composable reads from the i18n catalog (P1/S10) — the three
 * `costAnalysis.charts.*` keys the web component resolves via `t(...)` plus the accessible chrome the
 * native ChartContainer needs (a chart description and the fallback-table headers, drawn from existing
 * catalog keys since the web canvas relies on Recharts' implicit accessibility). The lifecycle-chrome
 * strings (error / retry / offline / freshness) are resolved inline at the Compose boundary, not here, so
 * this holder stays a thin content carrier.
 *
 * @property title the panel title (web `costAnalysis.charts.costPerKwh`).
 * @property rateLabel the line's tooltip name + the table rate-column header (web `costAnalysis.charts.rateLabel`).
 * @property noData the empty-state message (web `costAnalysis.charts.noData`).
 * @property accessibleDescription the chart's screen-reader description (the localized title).
 * @property dateColumn the accessible-table date-column header (the generic `common.date` key).
 */
data class CostPerKwhChartStrings(
    val title: String,
    val rateLabel: String,
    val noData: String,
    val accessibleDescription: String,
    val dateColumn: String,
)

/**
 * The fully projected, render-ready chart inputs — the native analogue of the web component's direct use
 * of its `data` prop plus the ChartContainer `data`/`dataColumns` accessible-table props. Pure data (no
 * Compose types) so the projection is unit-tested without a UI host: the composable wraps [values] into a
 * single `ChartSeries`, feeds [dates] to the line chart's bottom axis, and renders [tableRows] as the
 * accessible fallback table (`Date`, `$/kWh`).
 */
data class CostPerKwhChartProjectionResult(
    val dates: List<String>,
    val values: List<Double?>,
    val tableRows: List<List<String>>,
    val isEmpty: Boolean,
)

/**
 * The pure projection the composable renders — a 1:1 port of the web component's chart + table bindings.
 * Stateless and side-effect-free so it is fully covered by the off-device unit gate; the composable only
 * resolves localized strings, the palette color, and the freshness chrome.
 */
object CostPerKwhChartProjection {
    /**
     * Projects [points] into render-ready chart inputs, preserving input order (the parent already sorts
     * ascending by start time, so the component renders the series verbatim — the web never re-sorts).
     * [dates] feed the X axis (the web `<XAxis dataKey="date" />`), [values] become the single line
     * series, and each point contributes one accessible-table row (`[date, formatValue(costPerKwh)]`).
     * Injecting [formatValue] keeps the projection locale-deterministic for tests; the composable supplies
     * the real localized currency formatter. [isEmpty] mirrors the web `data.length > 0` empty branch.
     */
    fun project(
        points: List<CostPerKwhPoint>,
        formatValue: (cost: Double) -> String,
    ): CostPerKwhChartProjectionResult =
        CostPerKwhChartProjectionResult(
            dates = points.map { it.date },
            values = points.map { it.costPerKwh },
            tableRows = points.map { listOf(it.date, formatValue(it.costPerKwh)) },
            isEmpty = points.isEmpty(),
        )

    /**
     * Projects the web component's `{ data }` prop onto the shared cache-then-network [UiState], mirroring
     * the web component's only branch (`data.length > 0`): a non-empty series is [UiPhase.Content]; an
     * empty (or, defensively, `null`) series is [UiPhase.Empty] — the "Not enough data" surface. The web
     * component itself carries no loading or error branch (its parent gates those with a skeleton), so
     * neither does this overload projection; the loading / error / stale / offline states are driven only
     * through the stateful entry's host-owned [UiState].
     */
    fun projectUiState(points: List<CostPerKwhPoint>?): UiState<List<CostPerKwhPoint>> =
        if (points.isNullOrEmpty()) {
            UiState(phase = UiPhase.Empty, data = emptyList())
        } else {
            UiState(phase = UiPhase.Content, data = points)
        }

    /**
     * Locale-aware currency formatting — the native mirror of the web `useFormatting` `formatCurrency`
     * (`currencySymbol + fmtNumber(amount, decimals)`). A blank [symbol] degrades to [DEFAULT_CURRENCY]
     * (web `'$'`); a non-finite [amount] is coerced to 0 before formatting (web `safeNumber`), so the Y
     * axis and table never render `NaN`.
     */
    fun formatCurrency(
        amount: Double,
        symbol: String,
        decimals: Int = COST_DECIMALS,
        locale: Locale = Locale.getDefault(),
    ): String = "${symbol.ifBlank { DEFAULT_CURRENCY }}${ChartFormat.number(safeValue(amount), decimals.coerceAtLeast(0), locale)}"

    /** Coerces a non-finite value to 0 — the web `safeNumber` guard for formatter inputs. */
    internal fun safeValue(value: Double): Double = if (value.isFinite()) value else 0.0
}

/**
 * Emits the one PII-safe `view.opened` diagnostic with the surface [CostPerKwhChartRegistration.SLUG]
 * (P1/S11). Carries only the slug — never a date or cost — so a diagnostics line can never leak the
 * fleet's charging spend. Kept free of Compose so it is unit-tested with a recording [Logger]; the
 * composable calls it from its first-composition effect.
 */
fun recordCostPerKwhChartOpened(logger: Logger) {
    logger.info("view.opened", mapOf("surface" to CostPerKwhChartRegistration.SLUG))
}
